Persist widgets as screenshots to the notebook json.

jdfreder opened the pull request:
Finally!  This PR introduces a work around for the inability to render widgets on GitHub, nbviewer, through nbconvert, etc...  A save hook is used to persist canvas renderings of the widget DOM elements to the notebook JSON prior to save.

Ping @SylvainCorlay , @ellisonbg, @captainsafia, @willingc (we can use this for docs), @blink1073 (similar to your export button in the matplotlib widget renderer)

SylvainCorlay commented:
This is a great idea. I did not know about html2canvas! Does it handle svg?

jdfreder commented:
Yes, it should!

SylvainCorlay commented:
:tada:

jdfreder commented:
I should mention, this is WIP.

rgbkrk commented on ipywidgets/static/widgets/js/manager.js:
```diff
@@ -268,6 +288,96 @@ define([
+                
+                // Create a mime bundle for the screenshot. Remove
+                // URL information, so only the b64 encoded data 
+                // exists, because that's what the notebook likes.
```
and we should continue to question this assumption in future versions of both the message spec and the notebook

rgbkrk commented:
This is great. The on-save is a really neat idea.

ellisonbg commented:
Wow, does `html2canvas` work in the browser - IOW the live notebook - or just in node?

rgbkrk commented:
@ellisonbg Browserify (via wzrd.in at least) is certainly barfing on it, haven't tried anything else. If so :unamused:

willingc commented:
@jdfreder If you are at Cal Poly later this week, I would love to see it live. :surfer:

rgbkrk commented:
@willingc See https://gist.github.com/jdfreder/4909058126124ab46279

jdfreder commented:
> Wow, does html2canvas work in the browser - IOW the live notebook - or just in node

It works in the browser.



> @ellisonbg Browserify (via wzrd.in at least) is certainly barfing on it, haven't tried webpack. If so,

Not sure how you are importing it, but be aware it isn't an AMD module, it just pollutes the window with a `html2canvas` class.

rgbkrk commented:
Ok, cool. I assume they're just not providing commonjs then. I went to tinker and provide a link to a working sample to Brian, gave up at stack trace. Sorry for the alarm!

SylvainCorlay commented:
This solves the issue of RTD and GitHub rendering of notebooks which is pretty awesome.

Regarding nbviewer, I think that it should actually really support interactive widgets. 
After all, if it is ok to add a screenshot of the rendered widget in the notebook, it should also be ok to add information about the widget state which hopefully should be much smaller, so that widgets can be rendered dynamically on the page.

jdfreder commented:
Talked to @ellisonbg , and he suggested adding a render button for doing the HTML -> PNG conversion, since it is asynchronous.  I didn't know what icon to use, so I used a truck:

![screen shot 2016-01-13 at 3 19 11 pm](https://cloud.githubusercontent.com/assets/3292874/12311034/42ff6094-ba09-11e5-89f5-21aadb017784.png)

Here it is. in action:

![demo](https://cloud.githubusercontent.com/assets/3292874/12311057/77cc2bea-ba09-11e5-83bc-2ccb8afa51de.gif)

Here's the output, when saved:

https://gist.github.com/jdfreder/067476b62ed0f0042f89

Open to suggestions for the icon, otherwise, this is good to go!

SylvainCorlay commented:
I already use an auto for `auto-run` :)

SylvainCorlay commented:
@jdfreder the problem you have with save event handler that have to be fully synchronous is also going to arise if you want to save the widget state when the notebook is saved. 

It was not the case before, but since we use (asynchronously loaded) custom serializers, we will have the same issue.

Hence, it is probably something we should tackle at the notebook level.

jdfreder commented:
> Hence, it is probably something we should tackle in the notebook project.

Yeah I agree, but for now, I think this should be a button.  Then, when that is fixed in the notebook project, I can open an accompanying PR here to do it on save.  However, one thing I discovered is that DOM -> PNG can be really slow with a lot of widgets, which may be annoying on save.

SylvainCorlay commented on ipywidgets/static/widgets/js/manager.js:
```diff
@@ -141,7 +170,7 @@ define([
          * load_callback: function()
          *      function that is called when the widget manager state should be
-         *      loaded.  This function should return a promise for the widget
+         *      loaded.  This function should return a prooutput_metadatamise for the widget
```
?

jdfreder replied:
Thanks Atom!

minrk commented:
I don't think this should be in the toolbar, just the menubar (menubar should always be a superset of toolbar actions, anyway). Menubar space seems too precious for something like this.

jdfreder commented:
> don't think this should be in the toolbar, just the Menu (Menu should always be a superset of toolbar actions, anyway). Toolbar space seems too precious for something like this.
> 
> EDIT: Fixed super-confusing mixture of menubar/toolbar.

A menu is a really good idea, because in the not-so-distant-future we'll need a button for computing possible widget states.  Maybe we'll have more things too in the menu, related to managing widget state.

jdfreder commented:
Menu , done!

![demo](https://cloud.githubusercontent.com/assets/3292874/12363772/6c1a6966-bb7f-11e5-954b-e4c0eeb83f27.gif)

ellisonbg commented:
Looks like some bugs with the rendering of some widgets...

![screen shot 2016-01-15 at 2 06 59 pm](https://cloud.githubusercontent.com/assets/27600/12366652/52741a68-bb91-11e5-9a84-45cc981c7084.png)

jdfreder commented:
@ellisonbg thanks for testing this.  I think there may be problems with the library I've selected to do the renderings of the widgets.  I'll look into alternatives

jdfreder commented:
I investigated some more, part of the problem was the overflow: scroll attribute of the #site div in the notebook.  Temporarily removing that during the rendering seems to work well. Radio buttons, shadows, and select controls styled as a list box do not render.

jdfreder commented:
Okay, some good news.  I update the html2canvas component to the latest version and a lot of the bugs in the renderings are gone! Radio buttons and checkboxes now render correctly, see: https://gist.github.com/jdfreder/a2773c690bc84f211e15

List boxes still have a bug where only the selected element is rendered, but I've determined this is a bug in html2canvas and will open an issue there.

ellisonbg commented:
Sounds like progress!

On Wed, Jan 20, 2016 at 1:49 PM, Jonathan Frederic <notifications@github.com
> wrote:

> Okay, some good news. I update the html2canvas component to the latest
> version and a lot of the bugs in the renderings are gone! Radio buttons and
> checkboxes now render correctly, see:
> https://gist.github.com/jdfreder/a2773c690bc84f211e15
>
> List boxes still have a bug where only the selected element is rendered,
> but I've determined this is a bug in html2canvas and will open an issue
> there.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ipython/ipywidgets/pull/314#issuecomment-173371674>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

jdfreder commented:
> Sounds like progress!

It is, good progress.  The rendering fixes added in the latest version seemed to have caused a slowdown in the rendering speed, significantly.  It totally makes sense to have this in a menu or triggered by a button now.  I'm going to add a progress bar that allows you to know how long the render will take.

jdfreder commented:
Here's what this looks like with a nice loading screen.  The loading screen is full screen because scrolling in the notebook needs to be temporarily disabled as the widgets are rendered.

![loading](https://cloud.githubusercontent.com/assets/3292874/12468798/9236c94e-bf9b-11e5-8168-66eb6863aa3c.gif)

Thinking about this more, it probably makes sense for the menu option to be "Save with widgets".  The function will then automatically invoke a save after the widgets have been rasterized.  Saving using the normal save mechanism wouldn't fave with widget images.  

That way, the user would never accidentally save the notebook with old widget images, nor would the user have to perform two actions to save with widgets.  Opinions?

jasongrout commented:
just checking - to get the real widgets back, I would have to just re-execute everything again, and the images go away, and real widgets are there again?

jdfreder commented:
> just checking - to get the real widgets back, I would have to just re-execute everything again, and the images go away, and real widgets are there again?

No, actually the way it works now is that a custom metadata tag `metadata: {isWidgetSnapshot: true}` is added to the widget image outputs.  The widget Javascript, when loaded on the page, makes sure that those outputs are not visible.  Generating the snapshots leaves the live widgets as-is.

jdfreder commented:
I'm totally mystified as to why the Python tests fail in 0b2ff80 , seeing that I don't modify any Python in this PR.

@SylvainCorlay I rebased, added the download state action to the menu, and now save immediately after snapshot capture.  This is ready for review.  If you decide to test it, don't forget to rebuild your CSS!

jdfreder commented:
Demo:  

![loading](https://cloud.githubusercontent.com/assets/3292874/12518197/f040810a-c0eb-11e5-95b6-f4ee00802087.gif)

SylvainCorlay commented on ipywidgets/static/notebook/js/manager.js:
```diff
@@ -5,10 +5,16 @@ define([
-    "jupyter-js-widgets"
-], function (_, Backbone, comm, widgets) {
+    "jupyter-js-widgets",
+    "../../components/html2canvas/dist/html2canvas",
```
as soon as we separate jupyter-js-widgets, we should make those npm deps.

jdfreder replied:
I added a comment so we don't forget

jdfreder commented on ipywidgets/static/notebook/js/manager.js:
```diff
@@ -108,6 +121,20 @@ define([
+        // loaded from disk, and when the widget manager is constructed.
+        this.notebook.events.on('notebook_saved.Notebook', this.deleteSnapshots.bind(this));
+        this.notebook.events.on('notebook_save_failed.Notebook', this.deleteSnapshots.bind(this));
+        this.notebook.events.on('notebook_loaaded.Notebook', this.deleteSnapshots.bind(this));
```
loaaaaaaaded

jdfreder replied:
:heavy_check_mark:

SylvainCorlay commented:
A minor problem. You should probably disable scrolling while rendering the snapshots.

jdfreder commented:
> A minor problem. You should probably disable scrolling of the notebook while rendering the snapshots.

Last commit hides the scrollbars and fixes the scroll position.

SylvainCorlay commented:
:+1:
